compiler/x/amd64/codegen/OMRMemoryReference.cpp

```diff
@@ -507,6 +510,20 @@ OMR::X86::AMD64::MemoryReference::addMetaDataForCodeAddressWithLoad(
                                               containingInstruction->getNode(),
                                               counter);
          }
+      else if (sr.getSymbol()->isBlockFrequency())
+         {
+         if (cg->needRelocationsForPersistentProfileInfoData())
+            {
+            TR_RelocationRecordInformation *recordInfo = ( TR_RelocationRecordInformation *)cg->comp()->trMemory()->allocateMemory(sizeof( TR_RelocationRecordInformation), heapAlloc);
+            recordInfo->data1 = (uintptr_t)&sr;
+            recordInfo->data2 = 0; // seqKind
+            //TR::StaticSymbol *staticSym = sr.getSymbol()->getStaticSymbol();
+            //TR_ASSERT(staticSym, "Expected static symbol for block frequency\n");
+            //TR::Relocation *relocation = new (cg->trHeapMemory()) TR::ExternalRelocation(displacementLocation, (uint8_t *)staticSym->getStaticAddress(), TR_BlockFrequency, cg);
+            TR::Relocation *relocation = new (cg->trHeapMemory()) TR::ExternalRelocation(displacementLocation, (uint8_t *)recordInfo, TR_BlockFrequency, cg);
+            cg->addExternalRelocation(relocation, __FILE__, __LINE__, containingInstruction->getNode());
+            }
+         }
```

Comment:
How come you don't also need a relo record for `TR_RecompQueuedFlag` here?

Reply:
Looking at the relocation records generated with this change set I see `TR_BlockFrequency` is getting generated twice for the same offset, eg:
```
TR_BlockFrequency (102)             /home/asmehra/data/ashu-mehra/openj9-openjdk-jdk11/omr/compiler/x/amd64/codegen/OMRMemoryReference.cpp   524      0191       0151 00007F9E174BFD80
TargetAddress1:00007F9E15DE4D10,  TargetAddress2:0000000000000000
TR_BlockFrequency (102)             /home/asmehra/data/ashu-mehra/openj9-openjdk-jdk11/omr/compiler/x/codegen/X86BinaryEncoding.cpp  2937      0191       0151 00007F9E174BFD80
TargetAddress1:00007F9E15DE4DA0,  TargetAddress2:0000000000000000
```

but `TR_RecompQueuedFlag` is generated only once.

`OMR::X86::AMD64::MemoryReference::generateBinaryEncoding` creates a new instruction to move the addr to reg [1].
Then we add metadata for the address as part of which relocation records are generated [2].
This is the place where this patch has code for creating relo record for `TR_BlockFrequency` but not for `TR_RecompQueuedFlag`.
Then further down in `OMR::X86::AMD64::MemoryReference::generateBinaryEncoding` when the binary encoding for the new instruction is generated [3], we again generate metadata [4] where this patch creates relo records for `TR_BlockFrequency` and `TR_RecompQueuedFlag`.
So effectively we end up creating duplicate relo records for `TR_BlockFrequency`. We should be generating the relo record only at one place, which I feel should be when adding metadata for the instruction. Going by that logic, the relo record generated at [2] for `TR_BlockFrequency` should be avoided.

On similar note, probably `TR_DebugCounter` also has similar issue of duplicate records getting generated.

@dsouzai your thoughts on this?

[1] https://github.com/eclipse/omr/blob/bded46caf73287830139853961ffacb96cc9758e/compiler/x/amd64/codegen/OMRMemoryReference.cpp#L612
[2] https://github.com/eclipse/omr/blob/bded46caf73287830139853961ffacb96cc9758e/compiler/x/amd64/codegen/OMRMemoryReference.cpp#L644
[3] https://github.com/eclipse/omr/blob/bded46caf73287830139853961ffacb96cc9758e/compiler/x/amd64/codegen/OMRMemoryReference.cpp#L664
[4] https://github.com/eclipse/omr/blob/bded46caf73287830139853961ffacb96cc9758e/compiler/x/codegen/X86BinaryEncoding.cpp#L2796

Reply:
Hm interesting. Yeah I think you're right; the debug counter relo doesn't look like it belongs in [2] above, especially since there's no `TR_DataAddress` in [2]. So `TR_BlockFrequency` doesn't belong in [2] either.